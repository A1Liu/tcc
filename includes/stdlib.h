#define NULL ((void *)0)

typedef unsigned long size_t;

void *malloc(size_t size);
void *realloc(void *buffer, size_t new_size);
void free(void *value);
