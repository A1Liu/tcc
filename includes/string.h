
typedef unsigned long size_t;

void *memcpy(void *dest, void *src, size_t n);
size_t strlen(char *str);
