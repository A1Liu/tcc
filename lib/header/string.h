typedef unsigned long size_t;

void *memset(void *dest, int ch, size_t count);

void bcopy(const void *src, void *dest, size_t len);

void *memcpy(void *dest, void *src, size_t n);

size_t strlen(char *str);
