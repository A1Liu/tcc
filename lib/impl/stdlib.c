#include <errno.h>
#include <stdlib.h>
#include <tci.h>

void *malloc(size_t size) { return tci_ecall(TCI_ECALL_HEAP_ALLOC, size); }

void *realloc(void *_buffer, size_t new_size) {
  char *buffer = _buffer;

  size_t allocation_offset = 0;
  for (; tci_ecall(TCI_ECALL_IS_SAFE, buffer - 1 - allocation_offset);
       allocation_offset++)
    ;

  if (allocation_offset > 0)
    tci_throw_error(
        "InvalidPointer",
        "called realloc on a pointer in the middle of an allocation", 1);

  size_t old_size = tci_var_size(buffer);
  char *new_buffer = malloc(new_size);
  size_t copy_bound = new_size < old_size ? new_size : old_size;

  for (size_t i = 0; i < copy_bound; i++)
    new_buffer[i] = buffer[i];

  free(buffer);

  return new_buffer;
}

void free(void *buffer) {}

void exit(int status) { tci_ecall(TCI_ECALL_EXIT, status); }

long strtol(const char *restrict str, char **restrict str_end, int base) {}

unsigned long strtoul(const char *restrict str, char **restrict str_end,
                      int base) {}
