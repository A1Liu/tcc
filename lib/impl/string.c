#include <stdint.h>
#include <string.h>

void bcopy(const void *src, void *dest, size_t len) {
  if (dest < src) {
    const char *firsts = (const char *)src;
    char *firstd = (char *)dest;
    while (len--)
      *firstd++ = *firsts++;
  } else {
    const char *lasts = (const char *)src + (len - 1);
    char *lastd = (char *)dest + (len - 1);
    while (len--)
      *lastd-- = *lasts--;
  }
}

void *memcpy(void *_dest, void *_src, size_t n) {
  char *src = (char *)_src;
  char *dest = (char *)_dest;

  for (int i = 0; i < n; i++)
    dest[i] = src[i];

  return dest;
}

void *memset(void *s, int c, size_t n) {
  unsigned char *p = s;
  while (n--)
    *p++ = (unsigned char)c;
  return s;
}

size_t strlen(char *str) {
  size_t len = 0;
  for (; *str; len++, str = str + 1)
    ;

  return len;
}

int strcmp(const char *left, const char *right) {
  for (; *left && !(*left - *right); left++, right++)
    ;

  return *left - *right;
}
