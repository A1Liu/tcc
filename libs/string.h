#include <stdint.h>
#include <string.h>

void *memcpy(void *_dest, void *_src, size_t n) {
  char *src = (char *)_src;
  char *dest = (char *)_dest;

  for (int i = 0; i < n; i++)
    dest[i] = src[i];

  return dest;
}

size_t strlen(char *str) {
  size_t len = 0;
  for (; *str; len++, str = str + 1)
    ;

  return len;
}
