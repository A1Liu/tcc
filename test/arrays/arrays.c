#include "arrays.h"
#include <stdlib.h>
#include <string.h>

String string_new(char *value) {
  int len = strlen(value);
  String str = malloc(sizeof(int) + sizeof(char) * len + 1);

  for (int i = 0; i < len; i++) {
    str->data[i] = value[i];
  }
  str->data[len] = '\0';
  str->size = len;

  return str;
}
