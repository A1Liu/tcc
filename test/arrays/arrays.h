typedef struct {
  int size;
  char data[];
} * String;

String string_new(char *);
