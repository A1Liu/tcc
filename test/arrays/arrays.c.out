String of length 7: hello!!
